lib/src/dense_set.rs

```diff
@@ -0,0 +1,1282 @@
+        Self {
+            data: d,
+            reg_vec: Vec::new(),
+        }
+    }
+
+    fn unit(item: Reg) -> Self {
+        let mut s: BitSet = RegSet::empty();
+        s.insert(item);
+        s
+    }
+
+    fn two(item1: Reg, item2: Reg) -> Self {
+        let mut s: BitSet = RegSet::empty();
+        s.insert(item1);
+        s.insert(item2);
+        s
+    }
+
+    fn card(&self) -> usize {
+        let mut counter = 0;
+
+        for i in 0..self.data.len() {
+            counter += self.data[i].count_ones();
+        }
+        counter as usize
+    }
+
+    fn insert(&mut self, item: Reg) {
+        if !self.contains(item) {
+            self.data_insert(item);
+            self.vec_insert(item);
+        }
+    }
+
+    fn data_insert(&mut self, item: Reg) {
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        if data_index >= self.data.len() {
+            for _ in self.data.len()..data_index + 1 {
+                self.data.push(0);
+            }
+        }
+        self.data[data_index] |= 1 << offset;
+    }
+    fn vec_insert(&mut self, item: Reg) {
+        self.reg_vec.push(item);
+    }
+
+    fn contains(&self, item: Reg) -> bool {
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        if data_index >= self.data.len() {
+            false
+        } else {
+            (1 & (self.data[data_index] >> offset)) != 0
+        }
+    }
+
+    fn from_vec(vec: Vec<Reg>) -> Self {
+        let mut res: BitSet = RegSet::empty();
+
+        for &reg in vec.iter() {
+            res.data_insert(reg);
+        }
+        res.reg_vec = vec;
+
+        res
+    }
+
+    fn to_vec(&self) -> Vec<Reg> {
+        return self.reg_vec.clone();
+    }
+
+    fn delete(&mut self, item: Reg) {
+        self.vec_delete(item);
+        self.data_delete(item);
+    }
+
+    fn vec_delete(&mut self, item: Reg) {
+        let mut reg_index = 0;
+        for &mut reg in self.reg_vec.iter_mut() {
+            if reg == item {
+                self.reg_vec.swap_remove(reg_index);
+                break;
+            }
+            reg_index += 1;
+        }
+    }
+
+    fn data_delete(&mut self, item: Reg) {
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        self.data[data_index] &= !(1 << offset);
+    }
+
+    fn is_empty(&self) -> bool {
+        for i in 0..self.data.len() {
+            if self.data[i] != 0 {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn intersect(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+        let mut res = Vec::<Reg>::new();
+
+        for &item in self.reg_vec.iter() {
+            if other.contains(item) {
```

Comment:
This is very expensive: if there are N elements in self and M in other, it means we have N x M operations here, since `contains` also hides a linear operation.

I think there's a way to avoid this, by having `reg_vec` store the reg with index `i` at the index `i` in the `reg_vec`. So `reg_vec` would be a `Vec<Option<Reg>>`, with `None` indicating absence of the reg in set. It will consume more memory, but then, this test becomes trivial, there's no need to create a new replacement vector for `self.reg_vec` (you could mutate it in place) and the overall function becomes linear instead of being quadratic.

Of course, this is only interesting to implement if this is an actual issue. One thing would be to look at a profile created by `perf`, and see how high this function shows in a profile report. Or you could just try implementing it and see if it helps wallclock performance.

Reply:
Hello @bnjbvr 

Thanks for the review !

> it means we have N x M operations here, since contains also hides a linear operation.

Sorry, I don't see where it hides a linear operation (access `Vec` values by indexing seems O(1)):

```
    fn contains(&self, item: Reg) -> bool {
        let reg_index = BitSet::get_reg_index(item);
        let data_index = BitSet::get_data_index(reg_index);

        if data_index >= self.data.len() {
            false
        } else {
            let offset = BitSet::get_offset(reg_index);
            (1 & (self.data[data_index] >> offset)) != 0
        }
    }

    fn get_reg_index(item: Reg) -> usize {
        if item.is_real() {
            item.get_index()
        } else {
            NUMBER_REAL_REG + item.get_index()
        }
    }
    
    fn get_offset(reg_index: usize) -> usize {
        reg_index % BLOCK_SIZE
    }
    
    fn get_data_index(reg_index: usize) -> usize {
        reg_index / BLOCK_SIZE
    }

```

Reply:
> I think there's a way to avoid this, by having reg_vec store the reg with index i at the index i in the reg_vec. So reg_vec would be a Vec<Option<Reg>>, with None indicating absence of the reg in set. It will consume more memory, but then, this test becomes trivial, there's no need to create a new replacement vector for self.reg_vec (you could mutate it in place) and the overall function becomes linear instead of being quadratic.
> 
> Of course, this is only interesting to implement if this is an actual issue. One thing would be to look at a profile created by perf, and see how high this function shows in a profile report. Or you could just try implementing it and see if it helps wallclock performance.

I implemented it ff39c7c93a66c9ecced9bf607a1935b38ab31c51.

Before the commit:
```
   Total     Self  Pass
-------- --------  ----------------------------------
   0.016    0.000  Translate WASM module
   0.016    0.016  Translate WASM function
   0.043    0.040  Verify Cranelift IR
   0.000    0.000  Verify CPU flags
   0.645    0.051  Compilation passes
   0.002    0.002  Control flow graph
   0.002    0.002  Dominator tree
   0.000    0.000  Remove unreachable blocks
   0.550    0.550  Register allocation
   0.001    0.001  Binary machine code emission
```
After:

```
   Total     Self  Pass
-------- --------  ----------------------------------
   0.016    0.000  Translate WASM module
   0.016    0.016  Translate WASM function
   0.042    0.039  Verify Cranelift IR
   0.000    0.000  Verify CPU flags
   1.738    0.062  Compilation passes
   0.002    0.002  Control flow graph
   0.002    0.002  Dominator tree
   0.000    0.000  Remove unreachable blocks
   1.632    1.632  Register allocation
   0.001    0.001  Binary machine code emission
```